### ISSUE TITLE
About Runtime vs. Compiletime Checks

### ISSUE THREAD

**Issue by yakagami** [dart-lang/language#3277]
From the docs:

>The Dart language is type safe: it uses a combination of static type checking and [runtime checks](https://dart.dev/language/type-system#runtime-checks) to ensure that a variable’s value always matches the variable’s static type, sometimes referred to as sound typing.

There is an example of a runtime check there:

```dart
void main() {
  List<Animal> animals = [Dog()];
  List<Cat> cats = animals as List<Cat>; <--- runtime error
}  
```
My question is what amount of the type checking typically happens at compile time vs. runtime? I understand from  [Dart and the performance benefits of sound types](https://medium.com/dartlang/dart-and-the-performance-benefits-of-sound-types-6ceedd5b6cdc) that before null safety, dart did a lot of runtime checks because of null. 

Also from that link:

>With soundness, our tools guarantee that types are correct via a combination of static and (when needed) runtime checking. Without soundness, type checking can only go so far, and static types may be incorrect at runtime.

There can still seemingly be runtime type errors:

```dart
void main() {
    f({"x":3});
}
int f(Map<String, dynamic> o){
  g(o);
  return o["x"];
}
void g(o){
  o["x"] = "foo";
}
```
gives

>Unhandled exception:
>type 'String' is not a subtype of type 'int'

I modified this from the following code from a [TC39 JavaScript proposal (page 25)](https://docs.google.com/presentation/d/1OraKn6TUdmtug-lkgijq-43Xou_1lB1_DJ54x6ssnRY/edit#slide=id.g213cd9595b0_2_12) (see also [this accompanying text (paragraph 9)](https://github.com/tc39/notes/blob/main/meetings/2023-03/mar-22.md#type-annotations-proposal-update))
```ts
function f(o : { x:number }) : number {
  g(o); return o.x;
}
function g(o) { o.x = “foo”; }  // ERROR!
f({x:3});
```

I couldn't recreate the exact code in dart as I believe function parameters cannot demand specific string keys in dart ie. ` int f(Map<"x", dynamic> o){...}`.

Is this a runtime check? I've also seen runtime type errors here and there in flutter even with dart 3.0. So my question is: how expensive are runtime checks and how often are they performed? Is it by every function return statement? Or only for mutable objects like `Map<k,v>`?

Also if there are any resources to learn more about types and it's relationship to performance in general I would be interested. Thanks.

**Comment by munificent** [comment]
It's very hard to give blanket answers to questions like this because:

* The amount of time a given program spends in code doing runtime type checks depends wildly on the program itself. You can write programs in styles that never touch a single runtime check, or write programs that hit them on practically every assignment and method call.

* The actual performance cost of these checks depends heavily on the specific compiler and runtime implementation. Compiler and VM implementers spend a *lot* of time coming up with very creative ways to implement these checks quickly, so the cost is an often-changing moving target.

* Compilers can often eliminate many of these checks in places where they can prove they aren't needed. This is also a moving target.

So, in general, you tend to not get very concrete answers to this question. The answer is really "it depends". What we *can* do is clarify *where* these runtime checks happen in Dart code. Off the top of my head, I think the main ones are:

* Evaluating an `as` expression.
* Invoking any member on an object whose static type is `dynamic`.
* Assigning a value whose static type is `dynamic` to a place where another type is expected. (In other words, an implicit `as` cast from `dynamic` to another type.)
* Calling a method on a generic class whose parameter signature depends on type arguments. This is the most subtle one and is because Dart makes all generics covariant. For example:

    ```dart
    List<num> numbers = <int>[1, 2, 3];
    numbers.add(1.2); // A num, but not an int.
    ```

    This program fails at runtime because even though the call to `add()` in this example statically accepts any `num`, the actual list at runtime is a `List<int>` and `1.2` isn't an `int`. The runtime failure here is from the covariance checks implicitly inserted at the top of the `add()` method.

Obviously, `as` expressions have a runtime cost. If you avoid using `dynamic`, you avoid all of the runtime cost of the second two bullet points. The last one is basically impossible to avoid because generics are so widely used. That's why we're exploring [declaration-site variance](https://medium.com/dartlang/dart-declaration-site-variance-5c0e9c5f18a5).

**Comment by yakagami** [comment]
@munificent Thanks for the quick reply.

what about the following code:

```dart
void main() {
    f({"x":3});
}
int f(Map<String, int> o){
  g(o);
  return o["x"]!;
}
void g(o){
  o["x"] = "foo";
}
```

Is it the `!` that causes this? I was surprised that `!` was even needed here as the parameter for o is `Map<String, int>`. In any case this also is runtime checked even though all I declared is that `o` is not null (and not that it is an `int` as opposed to a `String`).

**Comment by lrhn** [comment]
> Is this a runtime check? 

Yes.

The runtime type checks in Dart are, generally:
* Any `is` operator (which isn't trivially true).
* Any `as` operator (which isn't trivially safe, fx, an up-cast is safe).
* Any `try`/`on` type check against a thrown object.
* Any type check which is part of a refutable pattern, like `case (Foo f)` or `case [var first, ...]` (checks if value is `List`).
* Any implicit downcast from `dynamic` (assigning something with static type `dynamic` to anything which isn't a top type introduces an implicit type check).
* Any argument passed to a covariant instance member parameter.
    * A covariant parameter is either declared as such `void foo(covariant int x) { ... }`, 
    * or implicitly so because it uses a class type parameter covariantly, like `void add(E element)` from `List<E>`.
* Any contravariant return type of an instance member, like reading `func` of a `class C<T> { void Function(T) get func => ...}`.
* Any `await e` where the static type of `e` is not definitely a `Future<T>` for some type `T`. 
    * That includes awaiting something with static type `FutureOr<T>`, or
    * awaiting something which is not a future type, like `Object` or `Foo` (which *might* have a subclass implementing `Future<Foo>`.
    * and returns in `async` functions also do an implicit await.
* Dynamic invocations of a member on `dynamic` or the `call` method of `Function`.
    * This does some dynamic member *lookup* first, which can also be costly, but then it also has to check that the actual arguments, and actual type arguments, have valid types for the member it finds.

How much time is taken by type checks depends on a lot of things, including which platform you're compiling for. 

Every one of these runtime type checks are there because they may go either way at runtime, and language semantics depend visibly on whether the check is true or false.
In many of these cases, it might be possible to predict some of the tests at runtime (doing `switch (aList) { case [var x, ...] => ...}` knows it doesn't need to check for `isList`). That's just optimization. Compilers do what they can to not do unnecessary work.

**Comment by yakagami** [comment]
@lrhn Thanks. How does the code here https://github.com/dart-lang/language/issues/3277#issuecomment-1679550091 work? It does not use `dynamic`. Is that one of the cases listed above?

**Comment by munificent** [comment]
Yes, I forgot about `!`, which is essentially like `as`. Your map isn't nullable, but the `[]` method on Map is defined to return `V?` because it returns `null` when the key is absent.

**Comment by yakagami** [comment]
So where is the runtime check happening? Does the compiler know that since `g` may have mutated `o` it can no longer rely on the parameter type? I don't see how the null check would invoke the compiler to add a runtime check that `o["x"]` is an int. Or are you saying that any `k,v` pair in `Map<String, int>` is essentially the same as `Map<String, dynamic>` because it could be null?

What I mean is, why does the compiler not rely on the parameter types here; obviously it shouldn't, but what is causing that to happen? Or does it never rely on the parameter type for `Map<String, T>`?

**Comment by munificent** [comment]
> So where is the runtime check happening?

The `!` operator *is* a runtime check. That expression evaluates its operand, checks if it is `null`, and throws an exception if so.

**Comment by yakagami** [comment]
ah. I thought this code should run print "runtime" before the runtime error, and the issue was the return statement not being of type `int`. So I assume there is 

```dart
void main() {
    print(f({"x":3}));
}

int f(Map<String, int> o){
  g(o); //<-- fails
  print("runtime");
  return o.values.first;
}
void g(o){
  o["x"] = "foo";
}
```

which *is* true for 

```dart
void main() {
    print(f({"x":3}));
}

int f(Map<String, dynamic> o){
  g(o);
  print("runtime"); // <-- prints
  return o.values.first; //<-- fails
}
void g(o){
  o["x"] = "foo";
}
```

Inlining g(o) like so gives a compiletime error:

void main() {
    f({"x":3});
}
int f(Map<String, int> o){
  o["x"] = "foo";
  return o.values.first;
}

>Error: A value of type 'String' can't be assigned to a variable of type 'int'.
>  o["x"] = "foo";

Is this a case of an "argument passed to a covariant instance member parameter"? Also I suppose in this trivial case an analyzer could have found the issue at compile time but in more complex cases that wouldn't be possible?

**Comment by munificent** [comment]
Here:

> ```dart
> void g(o){
>   o["x"] = "foo";
> }
> ```

Since `o` has no type annotation, it has type `dynamic`. That means all operations performed on it, including calling `[]=` will involve a runtime type test which may fail. In this case, it's trying to pass a string as the second argument to `[]=` but at runtime it discovers that the actual map object being called expects the parameter to be an int.
